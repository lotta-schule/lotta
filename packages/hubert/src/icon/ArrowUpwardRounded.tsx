import * as React from 'react';
import clsx from 'clsx';

import styles from './SvgIcon.module.scss';

export const ArrowUpwardRounded = React.memo<React.SVGProps<SVGSVGElement>>(
  ({ className, ...props }) => (
    <svg
      xmlns="http://www.w3.org/2000/svg"
      focusable="false"
      viewBox="0 0 24 24"
      aria-hidden="true"
      className={clsx(styles.root, className)}
      {...props}
    >
      <path d="M13 19V7.83l4.88 4.88c.39.39 1.03.39 1.42 0 .39-.39.39-1.02 0-1.41l-6.59-6.59a.9959.9959 0 0 0-1.41 0l-6.6 6.58c-.39.39-.39 1.02 0 1.41.39.39 1.02.39 1.41 0L11 7.83V19c0 .55.45 1 1 1s1-.45 1-1z" />
    </svg>
  )
);
ArrowUpwardRounded.displayName = 'ArrowUpwardRoundedIcon';
