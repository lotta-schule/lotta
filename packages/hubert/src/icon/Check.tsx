import * as React from 'react';
import clsx from 'clsx';

import styles from './SvgIcon.module.scss';

export const Check = React.memo<React.SVGProps<SVGSVGElement>>(
  ({ className, ...props }) => (
    <svg
      xmlns="http://www.w3.org/2000/svg"
      focusable="false"
      viewBox="0 0 24 24"
      aria-hidden="true"
      className={clsx(styles.root, className)}
      {...props}
    >
      <path d="M9 16.17 4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z" />
    </svg>
  )
);
Check.displayName = 'CheckIcon';
