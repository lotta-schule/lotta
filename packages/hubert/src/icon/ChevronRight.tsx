import * as React from 'react';
import clsx from 'clsx';

import styles from './SvgIcon.module.scss';

export const ChevronRight = React.memo<React.SVGProps<SVGSVGElement>>(
  ({ className, ...props }) => (
    <svg
      xmlns="http://www.w3.org/2000/svg"
      focusable="false"
      viewBox="0 0 24 24"
      aria-hidden="true"
      className={clsx(styles.root, className)}
      {...props}
    >
      <path d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z" />
    </svg>
  )
);
ChevronRight.displayName = 'ChevronRightIcon';
