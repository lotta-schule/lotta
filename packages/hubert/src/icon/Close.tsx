import * as React from 'react';
import clsx from 'clsx';

import styles from './SvgIcon.module.scss';

export const Close = React.memo<React.SVGProps<SVGSVGElement>>(
  ({ className, ...props }) => (
    <svg
      xmlns="http://www.w3.org/2000/svg"
      focusable="false"
      viewBox="0 0 24 24"
      aria-hidden="true"
      className={clsx(styles.root, className)}
      {...props}
    >
      <path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z" />
    </svg>
  )
);
Close.displayName = 'CloseIcon';
