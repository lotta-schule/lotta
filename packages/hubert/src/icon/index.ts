export * from './ArrowUpwardRounded';
export * from './Check';
export * from './ChevronRight';
export * from './Close';
export * from './DragHandle';
export * from './ExpandMore';
export * from './KeyboardArrowLeft';
export * from './KeyboardArrowRight';
