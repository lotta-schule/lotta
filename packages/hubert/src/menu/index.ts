export * from './MenuButton';
export { Item } from './MenuItem';
export * from './Menu';
