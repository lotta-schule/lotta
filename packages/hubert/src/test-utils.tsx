import * as React from 'react';
import { DefaultThemes } from '@lotta-schule/theme';
import { render, RenderOptions } from '@testing-library/react';
import { MotionConfig } from 'framer-motion';
import { HubertProvider } from './HubertProvider';
import { GlobalStyles } from './theme';

const theme = DefaultThemes.standard;

const Wrapper = ({ children }: any) => (
  <HubertProvider>
    <MotionConfig transition={{ duration: 0 }}>
      <GlobalStyles theme={theme} />
      {children}
    </MotionConfig>
  </HubertProvider>
);

const customRender = (
  ui: React.ReactElement,
  renderOptions: Omit<RenderOptions, 'wrapper'> = {}
) =>
  render(ui, {
    wrapper: Wrapper,
    ...renderOptions,
  });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
