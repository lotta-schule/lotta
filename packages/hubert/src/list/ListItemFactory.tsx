import * as React from 'react';
import { CollectionElement, Node } from '@react-types/shared';
import { Item } from '../menu';
import { ListItem, ListItemProps } from './ListItem';
import { Check } from '../icon';

export type ListItemPreliminaryItem = {
  key: React.Key;
  leftSection?: React.ReactNode;
  description?: string | React.ReactNode;
  rightSection?: React.ReactNode;
} & (
  | {
      label: string;
      textValue?: string;
    }
  | {
      label: React.ReactNode;
      textValue: string;
    }
) &
  (
    | {
        rightSection?: React.ReactNode;
        selected?: false;
      }
    | { selected?: true; rightSection?: React.ReactNode }
  );

export const ListItemFactory = {
  createItem: ({
    label,
    textValue,
    key,
    leftSection,
    description,
    ...other
  }: ListItemPreliminaryItem) => {
    const rightSection =
      'rightSection' in other && !!other.rightSection ? (
        other.rightSection
      ) : 'selected' in other && !!other.selected ? (
        <Check />
      ) : null;
    return (
      <Item key={key} textValue={textValue ?? (label as string)}>
        <span>{leftSection}</span>
        {label}
        {description ?? null}
        {rightSection}
      </Item>
    ) as CollectionElement<ListItemPreliminaryItem>;
  },

  createListItem: (
    item: Node<ListItemPreliminaryItem>,
    otherProps: ListItemProps & { ref?: React.Ref<HTMLLIElement> } = {}
  ) => {
    if (!Array.isArray(item.rendered)) {
      throw new Error('<Item /> must have more than one child');
    }
    const leftSection = item.rendered[0];
    const content = item.rendered[1];
    const description = item.rendered[2];
    const rightSection = item.rendered[3];

    return (
      <ListItem
        {...otherProps}
        leftSection={leftSection}
        rightSection={rightSection}
        isSelected={item.value?.selected ?? otherProps.selected}
      >
        <div>{content}</div>
        {description && <span>{description}</span>}
      </ListItem>
    );
  },
};
