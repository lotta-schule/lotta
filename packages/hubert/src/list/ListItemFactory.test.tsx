import * as React from 'react';
import { ItemProps } from '@react-types/shared';
import { Close, ChevronRight } from '../icon';
import { Item } from '../menu';
import { render } from '../test-utils';
import { ListItemFactory, ListItemPreliminaryItem } from './ListItemFactory';

describe('list/ListItemFactory', () => {
  describe('create Item', () => {
    it('should render an item with complete information for a listItem', () => {
      const preliminaryItem: ListItemPreliminaryItem = {
        key: 'test',
        label: 'Test',
        leftSection: <Close />,
        rightSection: <ChevronRight />,
        description: 'Test description',
      };

      const item = ListItemFactory.createItem(preliminaryItem);

      expect(item.key).toEqual('test');
      expect(item.type).toBe(Item);
      expect(
        (item.props as ItemProps<ListItemPreliminaryItem>).textValue
      ).toEqual('Test');
      expect(item.props.children).toHaveLength(4);
      expect(
        React.Children.toArray(item.props.children as any).find(
          (child: React.ReactNode) => child === 'Test'
        )
      ).toBeTruthy();
    });

    it('should also have four children for minimal information for a listItem', () => {
      const preliminaryItem: ListItemPreliminaryItem = {
        key: 'test',
        label: 'Test',
      };

      const item = ListItemFactory.createItem(preliminaryItem);

      expect(item.key).toEqual('test');
      expect(item.props.children).toHaveLength(4);
    });
  });

  describe('create ListItem', () => {
    it('should render a listItem with complete information from an item', () => {
      const screen = render(
        ListItemFactory.createListItem({
          rendered: ListItemFactory.createItem({
            key: 'test',
            label: 'Test',
            leftSection: <Close />,
            rightSection: <ChevronRight />,
            description: 'Test description',
          }).props.children,
        } as any)
      );
      expect(screen.getByRole('listitem')).toMatchInlineSnapshot(`
            <li
              class="li"
            >
              <div>
                <span>
                  <svg
                    aria-hidden="true"
                    class="root"
                    focusable="false"
                    viewBox="0 0 24 24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                    />
                  </svg>
                </span>
              </div>
              <div
                class="mainSection"
              >
                <div>
                  Test
                </div>
                <span>
                  Test description
                </span>
              </div>
              <div>
                <svg
                  aria-hidden="true"
                  class="root"
                  focusable="false"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                  />
                </svg>
              </div>
            </li>
      `);
    });

    it('should render a listItem with complete information from an item', () => {
      const screen = render(
        ListItemFactory.createListItem({
          rendered: ListItemFactory.createItem({
            key: 'test',
            label: 'Test',
          }).props.children,
        } as any)
      );
      expect(screen.getByRole('listitem')).toMatchInlineSnapshot(`
            <li
              class="li"
            >
              <div>
                <span />
              </div>
              <div
                class="mainSection"
              >
                <div>
                  Test
                </div>
              </div>
            </li>
      `);
    });
  });
});
