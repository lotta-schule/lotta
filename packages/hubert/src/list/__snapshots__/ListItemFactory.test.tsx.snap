// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`list/ListItemFactory > create Item > should render an item with complete information for a listItem 1`] = `"[{"type":"span","key":".0","ref":null,"props":{"children":{"type":{"compare":null},"key":null,"ref":null,"props":{},"_owner":null,"_store":{}}},"_owner":null,"_store":{}},{"type":"span","key":".1","ref":null,"props":{"children":"Test"},"_owner":null,"_store":{}},"Test description",{"type":{"compare":null},"key":".3","ref":null,"props":{},"_owner":null,"_store":{}}]"`;

exports[`list/ListItemFactory > create ListItem > should render a listItem with complete information from an item 1`] = `
<li
  class="_li_5a05a4"
>
  <div>
    <span>
      <svg
        aria-hidden="true"
        class="_root_2ade80"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
    </span>
  </div>
  <div
    class="_mainSection_5a05a4"
  >
    <div>
      <span>
        Test
      </span>
    </div>
    <span>
      Test description
    </span>
  </div>
  <div>
    <svg
      aria-hidden="true"
      class="_root_2ade80"
      focusable="false"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
      />
    </svg>
  </div>
</li>
`;

exports[`list/ListItemFactory > create ListItem > should render a listItem with complete information from an item 2`] = `
<li
  class="_li_5a05a4"
>
  <div>
    <span />
  </div>
  <div
    class="_mainSection_5a05a4"
  >
    <div>
      <span>
        Test
      </span>
    </div>
  </div>
</li>
`;
